### PR TITLE
Added streams to CSV reader and writer api

### PR DESCRIPTION
**Pull request by shrshi**
## Description

This PR contributes to https://github.com/rapidsai/cudf/issues/13744.
-Added stream parameters to public APIs
`cudf::io::read_csv`
`cudf::io::write_csv`
-Added stream gtests

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Review by karthikeyann — APPROVED** [review]

File: cpp/tests/streams/io/csv_test.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  static uniform_distribution dist{};
+  std::generate_n(values.begin(), size, [&]() { return T{dist(engine)}; });
+
+  return values;
+}
+
+TEST_F(CSVTest, CSVReader)
+{
+  constexpr auto num_rows = 10;
+  auto int8_values        = random_values<int8_t>(num_rows);
```
Comment (vuule) [inline cpp/tests/streams/io/csv_test.cpp:62]: I think this the same situation as with ORC, where we don't really need the random values.

File: cpp/tests/streams/io/csv_test.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      .dtypes({cudf::data_type{cudf::type_id::INT8},
+               cudf::data_type{cudf::type_id::INT16},
+               cudf::data_type{cudf::type_id::INT32},
+               cudf::data_type{cudf::type_id::INT64},
+               cudf::data_type{cudf::type_id::UINT8},
+               cudf::data_type{cudf::type_id::UINT16},
+               cudf::data_type{cudf::type_id::UINT32},
+               cudf::data_type{cudf::type_id::UINT64},
+               cudf::data_type{cudf::type_id::FLOAT32},
+               cudf::data_type{cudf::type_id::FLOAT64}});
```
Comment (vuule) [inline cpp/tests/streams/io/csv_test.cpp:98]: If we don't pass the data types the reader will launch an extra kernel to infer them, so we should not include this option :)

File: cpp/tests/streams/io/csv_test.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+{
+  constexpr auto num_rows = 10;
+  auto int8_values        = random_values<int8_t>(num_rows);
+  auto int16_values       = random_values<int16_t>(num_rows);
+  auto int32_values       = random_values<int32_t>(num_rows);
+  auto int64_values       = random_values<int64_t>(num_rows);
+  auto uint8_values       = random_values<uint8_t>(num_rows);
+  auto uint16_values      = random_values<uint16_t>(num_rows);
+  auto uint32_values      = random_values<uint32_t>(num_rows);
+  auto uint64_values      = random_values<uint64_t>(num_rows);
```
Comment (vuule) [inline cpp/tests/streams/io/csv_test.cpp:69]: AFAICT this can all be a single column, the code path is the same

File: cpp/tests/streams/io/csv_test.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  auto float32_values     = random_values<float>(num_rows);
+  auto float64_values     = random_values<double>(num_rows);
```
Comment (vuule) [inline cpp/tests/streams/io/csv_test.cpp:71]: would be good to add a string column, there are additional thrust calls that post-process quotes (IIRC).

**Review by vuule — CHANGES_REQUESTED** [review]
Nice and clean API change.
I think we can hit a few more code paths in the tests :)

**Comment by shrshi** [comment]
Thank you for the feedback, @vuule. Changing the tests did reveal more code paths creating string scalars using the default stream, which are now fixed.

File: cpp/src/io/csv/writer_impl.cu [hunk 5]
```diff
@@ -146,6 +146,12 @@ struct column_to_strings_fn {
   {
   }
 
+  ~column_to_strings_fn()                                      = default;
+  column_to_strings_fn(column_to_strings_fn const&)            = delete;
+  column_to_strings_fn& operator=(column_to_strings_fn const&) = delete;
+  column_to_strings_fn(column_to_strings_fn&&)                 = delete;
+  column_to_strings_fn& operator=(column_to_strings_fn&&)      = delete;
```
Comment (PointKernel) [inline cpp/src/io/csv/writer_impl.cu:153]: Any particular reason we explicitly disable them?

File: cpp/src/io/csv/writer_impl.cu [hunk 6]
```diff
@@ -367,10 +374,10 @@ void write_chunked(data_sink* out_sink,
 
   CUDF_EXPECTS(str_column_view.size() > 0, "Unexpected empty strings column.");
 
-  cudf::string_scalar newline{options.get_line_terminator()};
+  cudf::string_scalar newline(options.get_line_terminator(), true, stream);
```
Comment (PointKernel) [inline cpp/src/io/csv/writer_impl.cu:377]:
```suggestion
  cudf::string_scalar newline{options.get_line_terminator(), true, stream};
```
nit: curly brackets are preferred for initialization.
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es23-prefer-the--initializer-syntax

The same in other places.

File: cpp/src/io/csv/writer_impl.cu [hunk 7]
```diff
@@ -470,18 +479,20 @@ void write_csv(data_sink* out_sink,
       // concatenate columns in each row into one big string column
       // (using null representation and delimiter):
       //
-      std::string delimiter_str{options.get_inter_column_delimiter()};
+      // std::string delimiter_str{options.get_inter_column_delimiter()};
```
Comment (PointKernel) [inline cpp/src/io/csv/writer_impl.cu:482]:
```suggestion
```
probably a leftoever

File: cpp/src/io/csv/writer_impl.cu (hunk 5 above)
Comment (shrshi) [reply cpp/src/io/csv/writer_impl.cu:153]: The `column_to_strings_fn` is not meant to be copied (https://github.com/rapidsai/cudf/pull/14313#discussion_r1374549183)

Comment (PointKernel) [reply cpp/src/io/csv/writer_impl.cu:153]: Good to know! Thanks!

File: cpp/src/io/csv/writer_impl.cu (hunk 6 above)
Comment (shrshi) [reply cpp/src/io/csv/writer_impl.cu]: Thanks, addressed!

File: cpp/src/io/csv/writer_impl.cu (hunk 7 above)
Comment (shrshi) [reply cpp/src/io/csv/writer_impl.cu]: Thanks, addressed!

**Comment by PointKernel** [comment]
@shrshi you may want to resolve the merge conflicts in test CMake first to unblock CI.

File: cpp/src/io/csv/writer_impl.cu [hunk 8]
```diff
@@ -160,8 +166,9 @@ struct column_to_strings_fn {
   std::enable_if_t<std::is_same_v<column_type, bool>, std::unique_ptr<column>> operator()(
     column_view const& column) const
   {
-    return cudf::strings::detail::from_booleans(
-      column, options_.get_true_value(), options_.get_false_value(), stream_, mr_);
+    string_scalar true_string{options_.get_true_value(), true, stream_, mr_};
+    string_scalar false_string{options_.get_false_value(), true, stream_, mr_};
```
Comment (vuule) [inline cpp/src/io/csv/writer_impl.cu:170]: The user-provided memory resource should only be used for the buffers that are returned to the user. Internally we should always use the default mr.
```suggestion
    string_scalar true_string{options_.get_true_value(), true, stream_};
    string_scalar false_string{options_.get_false_value(), true, stream_};
```

File: cpp/src/io/csv/writer_impl.cu [hunk 9]
```diff
@@ -175,7 +182,7 @@ struct column_to_strings_fn {
     }
 
     // handle special characters: {delimiter, '\n', "} in row:
-    string_scalar delimiter{std::string{options_.get_inter_column_delimiter()}, true, stream_};
+    string_scalar delimiter{std::string{options_.get_inter_column_delimiter()}, true, stream_, mr_};
```
Comment (vuule) [inline cpp/src/io/csv/writer_impl.cu:185]: same here, delimiter is not returned to the user
```suggestion
    string_scalar delimiter{std::string{options_.get_inter_column_delimiter()}, true, stream_};
```

File: cpp/tests/streams/io/csv_test.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/default_stream.hpp>
+#include <cudf_test/iterator_utilities.hpp>
+
+#include <random>
```
Comment (vuule) [inline cpp/tests/streams/io/csv_test.cpp:28]: no longer needed, I assume?
```suggestion
```

**Review by vuule — CHANGES_REQUESTED** [review]
Thank you for tracking down the default stream use!
Couple more tiny changes

File: cpp/tests/streams/io/csv_test.cpp (hunk 10 above)
Comment (shrshi) [reply cpp/tests/streams/io/csv_test.cpp]: Yes, also didn't need `stringstream`

File: cpp/src/io/csv/writer_impl.cu (hunk 8 above)
Comment (shrshi) [reply cpp/src/io/csv/writer_impl.cu]: Thanks, good to know!

**Review by vuule — APPROVED** [review]

File: cpp/include/cudf_test/column_wrapper.hpp [hunk 11]
```diff
@@ -803,7 +803,8 @@ class strings_column_wrapper : public detail::column_wrapper {
       offsets, cudf::test::get_default_stream(), rmm::mr::get_current_device_resource());
     auto d_bitmask = cudf::detail::make_device_uvector_sync(
       null_mask, cudf::test::get_default_stream(), rmm::mr::get_current_device_resource());
-    wrapped = cudf::make_strings_column(d_chars, d_offsets, d_bitmask, null_count);
+    wrapped = cudf::make_strings_column(
+      d_chars, d_offsets, d_bitmask, null_count, cudf::test::get_default_stream());
```
Comment (ttnghia) [inline cpp/include/cudf_test/column_wrapper.hpp:807]: Does this include changes from other PR? The additional stream parameter here and below seems unrelated to this PR.

**Review by PointKernel — APPROVED** [review]
LGTM

Thanks!

File: cpp/include/cudf_test/column_wrapper.hpp (hunk 11 above)
Comment (vyasr) [reply cpp/include/cudf_test/column_wrapper.hpp:807]: My guess is that CSV stream test is exercising this code path while no prior stream test was, so before this PR we didn't observe that there was a missing stream argument here.

Comment (shrshi) [reply cpp/include/cudf_test/column_wrapper.hpp:807]: Yep, this change is needed for the string column in the tests - `make_strings_column` is called with the default stream otherwise.
